frontend/src/firebase/database.tsx

```diff
@@ -0,0 +1,249 @@
+import {
+  equalTo,
+  get,
+  orderByValue,
+  push,
+  query,
+  ref,
+  set,
+} from "firebase/database";
+import { LatLng } from "leaflet";
+import { useListVals, useObjectVal } from "react-firebase-hooks/database";
+import { db } from "./firebase";
+
+const GROUPS = "groups";
+const USERS = "users";
+const RIDES = "rides";
+const PASSENGERS = "passengers";
+const ROUTES = "routes";
+const KEY_SLUG_OPTS = {
+  replacement: "-",
+  remove: undefined,
+  lower: true,
+  strict: true,
+  locale: "en",
+  trim: true,
+};
+
+export const DBConstants = {
+  GROUPS,
+  USERS,
+  RIDES,
+  PASSENGERS,
+  ROUTES,
+  KEY_SLUG_OPTS,
+};
+
+export type Group = {
+  id: string;
+  name: string;
+  description: string;
+  isPrivate: boolean;
+  rides: { [key: string]: boolean };
+  members: { [key: string]: boolean };
+  owner: string;
+  banner?: string;
+};
+
+export type Vehicle = {
+  carId?: string;
+  type: string;
+  fuelUsage: number;
+  numSeats: number;
+  displayName?: string;
+};
+
+export type User = {
+  uid: string;
+  name: string;
+  authProvider: string;
+  email: string;
+  cars?: Vehicle[];
+};
+
+export type Ride = {
+  id?: string;
+  name: string;
+  start: { lat: number; lng: number };
+  end: { lat: number; lng: number };
+  driver?: string;
+  isComplete: boolean;
+  carId?: string;
+  startDate: string;
+  endDate: string;
+};
+
+export type Route = {
+  distance: number;
+  fuelUsed: number;
+  shape: LatLng[];
+};
+
+export const getGroup = async (groupId: string) => {
+  return new Promise<Group>((resolve, reject) => {
+    get(ref(db, `${GROUPS}/${groupId}`)).then(
+      (result) => {
+        if (result.exists()) {
+          const group = result.val();
+          group.id = groupId;
+          resolve(group);
+        } else {
+          reject(undefined);
+        }
+      },
+      (error) => {
+        reject(error);
+      }
+    );
+  });
+};
+
+export const useGroup = (groupId: string) => {
+  return useObjectVal<Group>(ref(db, `${GROUPS}/${groupId}`), {
+    keyField: "id",
+  });
+};
+
+export const useGroups = () => {
```

Comment:
Could we call this "useGroupList"? It's really easy to forget a "s" and end up with the wrong function.